File: ShuffleList.java
// ShuffleList.java - Перемешиватель списков на Java (CLI)
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.function.*;
import java.util.stream.*;

public class ShuffleList {
    private static final Scanner scanner = new Scanner(System.in);
    private static final Random random = new Random();

    // ========== АЛГОРИТМЫ ПЕРЕМЕШИВАНИЯ ==========
    interface ShuffleFunc<T> {
        List<T> shuffle(List<T> list, Random rng);
    }

    static <T> List<T> fisherYates(List<T> list, Random rng) {
        List<T> result = new ArrayList<>(list);
        for (int i = result.size() - 1; i > 0; i--) {
            int j = rng.nextInt(i + 1);
            Collections.swap(result, i, j);
        }
        return result;
    }

    static <T> List<T> sortKeyShuffle(List<T> list, Random rng) {
        List<T> result = new ArrayList<>(list);
        result.sort((a, b) -> rng.nextBoolean() ? -1 : 1);
        return result;
    }

    static <T> List<T> builtinShuffle(List<T> list, Random rng) {
        List<T> result = new ArrayList<>(list);
        Collections.shuffle(result, rng);
        return result;
    }

    static class Algorithm<T> {
        String name;
        ShuffleFunc<T> func;
        Algorithm(String name, ShuffleFunc<T> func) {
            this.name = name;
            this.func = func;
        }
    }

    private static final Map<String, Algorithm<Object>> ALGORITHMS = new LinkedHashMap<>();
    static {
        ALGORITHMS.put("1", new Algorithm<>("Fisher-Yates (оптимальный)", ShuffleList::fisherYates));
        ALGORITHMS.put("2", new Algorithm<>("Сортировка случайным ключом", ShuffleList::sortKeyShuffle));
        ALGORITHMS.put("3", new Algorithm<>("Встроенный (стандартный)", ShuffleList::builtinShuffle));
    }

    // ========== СТАТИСТИКА ==========
    static <T extends Comparable<T>> int countInversions(List<T> list) {
        int inv = 0;
        for (int i = 0; i < list.size(); i++) {
            for (int j = i + 1; j < list.size(); j++) {
                if (list.get(i).compareTo(list.get(j)) > 0) inv++;
            }
        }
        return inv;
    }

    static double entropy(List<Double> list) {
        if (list.isEmpty()) return 0.0;
        double min = list.stream().min(Double::compare).orElse(0.0);
        double max = list.stream().max(Double::compare).orElse(0.0);
        if (max == min) return 0.0;
        int bins = 10;
        int[] hist = new int[bins];
        for (double x : list) {
            int idx = (int)((x - min) / (max - min) * bins);
            if (idx == bins) idx = bins - 1;
            hist[idx]++;
        }
        double ent = 0.0;
        double n = list.size();
        for (int count : hist) {
            if (count > 0) {
                double p = count / n;
                ent -= p * (Math.log(p) / Math.log(2));
            }
        }
        return ent;
    }

    static <T> Map<String, Object> getStats(List<T> list) {
        Map<String, Object> stats = new HashMap<>();
        stats.put("length", list.size());
        if (!list.isEmpty() && list.get(0) instanceof Number) {
            List<Double> nums = list.stream().map(x -> ((Number)x).doubleValue()).collect(Collectors.toList());
            stats.put("inversions", countInversions((List<Comparable>) list));
            stats.put("entropy", entropy(nums));
        } else {
            stats.put("inversions", 0);
            stats.put("entropy", 0.0);
        }
        return stats;
    }

    // ========== ВВОД/ВЫВОД ==========
    static List<Object> parseList(String input) {
        String[] parts = input.split(",");
        List<Object> result = new ArrayList<>();
        for (String p : parts) {
            p = p.trim();
            if (p.isEmpty()) continue;
            try {
                int val = Integer.parseInt(p);
                result.add(val);
            } catch (NumberFormatException e) {
                try {
                    double val = Double.parseDouble(p);
                    result.add(val);
                } catch (NumberFormatException ex) {
                    result.add(p);
                }
            }
        }
        return result;
    }

    static String formatList(List<?> list) {
        return list.toString();
    }

    static void exportJSON(List<?> list, String filename) throws IOException {
        String json = new com.google.gson.GsonBuilder().setPrettyPrinting().create().toJson(list);
        Files.write(Paths.get(filename), json.getBytes());
    }

    static void exportCSV(List<?> list, String filename) throws IOException {
        try (PrintWriter pw = new PrintWriter(filename)) {
            pw.println("value");
            for (Object o : list) pw.println(o);
        }
    }

    static List<Object> importList(String filename) throws IOException {
        String content = new String(Files.readAllBytes(Paths.get(filename)));
        if (filename.endsWith(".json")) {
            // Используем Gson
            return new com.google.gson.Gson().fromJson(content, List.class);
        } else {
            // CSV
            String[] lines = content.split("\n");
            List<Object> result = new ArrayList<>();
            for (int i = 1; i < lines.length; i++) {
                String line = lines[i].trim();
                if (line.isEmpty()) continue;
                try { result.add(Double.parseDouble(line)); }
                catch (NumberFormatException e) { result.add(line); }
            }
            return result;
        }
    }

    static List<Object> generateRandomList(int size, int minVal, int maxVal) {
        List<Object> list = new ArrayList<>();
        Random rnd = new Random();
        for (int i = 0; i < size; i++) {
            list.add(rnd.nextInt(maxVal - minVal + 1) + minVal);
        }
        return list;
    }

    // ========== ОСНОВНАЯ ЛОГИКА ==========
    static void interactive() throws IOException {
        System.out.println("🎲 ПЕРЕМЕШИВАТЕЛЬ СПИСКОВ");
        System.out.println("Введите список (через запятую):");
        System.out.print("> ");
        String input = scanner.nextLine().trim();
        List<Object> data;
        if (input.isEmpty()) {
            System.out.print("Размер случайного списка (по умолчанию 10): ");
            String sizeStr = scanner.nextLine().trim();
            int size = sizeStr.isEmpty() ? 10 : Integer.parseInt(sizeStr);
            data = generateRandomList(size, 0, 100);
            System.out.println("Сгенерированный список: " + formatList(data));
        } else {
            data = parseList(input);
            System.out.println("Исходный список: " + formatList(data));
        }

        System.out.println("\nВыберите алгоритм:");
        for (Map.Entry<String, Algorithm<Object>> entry : ALGORITHMS.entrySet()) {
            System.out.println(entry.getKey() + ". " + entry.getValue().name);
        }
        System.out.print("Ваш выбор (по умолчанию 1): ");
        String algoChoice = scanner.nextLine().trim();
        if (algoChoice.isEmpty() || !ALGORITHMS.containsKey(algoChoice)) algoChoice = "1";
        Algorithm<Object> algo = ALGORITHMS.get(algoChoice);

        System.out.print("Введите seed (или оставьте пустым для случайного): ");
        String seedStr = scanner.nextLine().trim();
        Random rng;
        if (!seedStr.isEmpty()) {
            try {
                long seed = Long.parseLong(seedStr);
                rng = new Random(seed);
            } catch (NumberFormatException e) {
                rng = new Random();
            }
        } else {
            rng = new Random();
        }

        List<Object> shuffled = algo.func.shuffle(data, rng);
        System.out.println("\nПеремешанный список: " + formatList(shuffled));

        Map<String, Object> stats = getStats(shuffled);
        System.out.println("\n📊 Статистика:");
        System.out.println("  Длина: " + stats.get("length"));
        System.out.println("  Инверсий: " + stats.get("inversions"));
        System.out.println("  Энтропия: " + stats.get("entropy"));

        System.out.print("\nЭкспортировать результат? (y/n): ");
        String exportChoice = scanner.nextLine().trim().toLowerCase();
        if (exportChoice.equals("y")) {
            System.out.print("Формат (json/csv): ");
            String fmt = scanner.nextLine().trim();
            System.out.print("Имя файла: ");
            String filename = scanner.nextLine().trim();
            if (filename.isEmpty()) filename = "shuffled." + fmt;
            if (fmt.equals("json")) exportJSON(shuffled, filename);
            else exportCSV(shuffled, filename);
            System.out.println("Экспортировано в " + filename);
        }
    }

    static void cli(String[] args) throws IOException {
        String list = null, file = null, algorithm = "1", export = null;
        int randomSize = 0;
        long seed = 0;
        boolean statsFlag = false;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--list")) list = args[++i];
            else if (args[i].equals("--file")) file = args[++i];
            else if (args[i].equals("--random")) randomSize = Integer.parseInt(args[++i]);
            else if (args[i].equals("--algorithm")) algorithm = args[++i];
            else if (args[i].equals("--seed")) seed = Long.parseLong(args[++i]);
            else if (args[i].equals("--export")) export = args[++i];
            else if (args[i].equals("--stats")) statsFlag = true;
        }

        List<Object> data;
        if (list != null) {
            data = parseList(list);
        } else if (file != null) {
            data = importList(file);
        } else if (randomSize > 0) {
            data = generateRandomList(randomSize, 0, 100);
        } else {
            System.out.println("Укажите --list, --file или --random");
            return;
        }
        if (data.isEmpty()) {
            System.out.println("Список пуст.");
            return;
        }

        Random rng = (seed != 0) ? new Random(seed) : new Random();
        Algorithm<Object> algo = ALGORITHMS.getOrDefault(algorithm, ALGORITHMS.get("1"));
        List<Object> shuffled = algo.func.shuffle(data, rng);

        System.out.println("Исходный список: " + formatList(data));
        System.out.println("Перемешанный список: " + formatList(shuffled));
        if (statsFlag) {
            Map<String, Object> stats = getStats(shuffled);
            System.out.println("\n📊 Статистика:");
            System.out.println("  Длина: " + stats.get("length"));
            System.out.println("  Инверсий: " + stats.get("inversions"));
            System.out.println("  Энтропия: " + stats.get("entropy"));
        }
        if (export != null) {
            if (export.endsWith(".json")) exportJSON(shuffled, export);
            else exportCSV(shuffled, export);
            System.out.println("Экспортировано в " + export);
        }
    }

    public static void main(String[] args) throws IOException {
        if (args.length > 0) {
            cli(args);
        } else {
            interactive();
        }
    }
}

File: ShuffleList.cs
// ShuffleList.cs - Перемешиватель списков на C# (CLI)
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace ShuffleList
{
    class Program
    {
        private static readonly Random random = new Random();

        // ========== АЛГОРИТМЫ ПЕРЕМЕШИВАНИЯ ==========
        static List<T> FisherYates<T>(List<T> list, Random rng)
        {
            var result = new List<T>(list);
            for (int i = result.Count - 1; i > 0; i--)
            {
                int j = rng.Next(i + 1);
                (result[i], result[j]) = (result[j], result[i]);
            }
            return result;
        }

        static List<T> SortKeyShuffle<T>(List<T> list, Random rng)
        {
            var result = new List<T>(list);
            result.Sort((a, b) => rng.NextDouble() < 0.5 ? -1 : 1);
            return result;
        }

        static List<T> BuiltinShuffle<T>(List<T> list, Random rng)
        {
            var result = new List<T>(list);
            // Используем Fisher-Yates с нашим генератором
            return FisherYates(result, rng);
        }

        static Dictionary<string, (string name, Func<List<object>, Random, List<object>> func)> algorithms =
            new Dictionary<string, (string, Func<List<object>, Random, List<object>>)>
        {
            {"1", ("Fisher-Yates (оптимальный)", (list, rng) => FisherYates(list, rng).Cast<object>().ToList())},
            {"2", ("Сортировка случайным ключом", (list, rng) => SortKeyShuffle(list, rng).Cast<object>().ToList())},
            {"3", ("Встроенный (стандартный)", (list, rng) => BuiltinShuffle(list, rng).Cast<object>().ToList())}
        };

        // ========== СТАТИСТИКА ==========
        static int CountInversions<T>(List<T> list) where T : IComparable<T>
        {
            int inv = 0;
            for (int i = 0; i < list.Count; i++)
                for (int j = i + 1; j < list.Count; j++)
                    if (list[i].CompareTo(list[j]) > 0) inv++;
            return inv;
        }

        static double Entropy(List<double> list)
        {
            if (list.Count == 0) return 0;
            double min = list.Min();
            double max = list.Max();
            if (max == min) return 0;
            int bins = 10;
            int[] hist = new int[bins];
            foreach (double x in list)
            {
                int idx = (int)((x - min) / (max - min) * bins);
                if (idx == bins) idx = bins - 1;
                hist[idx]++;
            }
            double ent = 0;
            double n = list.Count;
            foreach (int count in hist)
                if (count > 0)
                {
                    double p = count / n;
                    ent -= p * Math.Log2(p);
                }
            return ent;
        }

        static (int length, int inversions, double entropy) GetStats(List<object> list)
        {
            int length = list.Count;
            int inv = 0;
            double ent = 0.0;
            if (length > 0 && list[0] is IComparable)
            {
                var comparableList = list.Cast<IComparable>().ToList();
                inv = CountInversions(comparableList);
            }
            if (length > 0 && list[0] is double || list[0] is int)
            {
                var nums = list.Select(x => Convert.ToDouble(x)).ToList();
                ent = Entropy(nums);
            }
            return (length, inv, ent);
        }

        // ========== ВВОД/ВЫВОД ==========
        static List<object> ParseList(string input)
        {
            var parts = input.Split(',', StringSplitOptions.RemoveEmptyEntries);
            var result = new List<object>();
            foreach (var p in parts)
            {
                string trimmed = p.Trim();
                if (int.TryParse(trimmed, out int i)) result.Add(i);
                else if (double.TryParse(trimmed, out double d)) result.Add(d);
                else result.Add(trimmed);
            }
            return result;
        }

        static string FormatList(List<object> list) => $"[{string.Join(", ", list)}]";

        static void ExportJSON(List<object> list, string filename)
        {
            string json = JsonSerializer.Serialize(list, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(filename, json);
        }

        static void ExportCSV(List<object> list, string filename)
        {
            using var sw = new StreamWriter(filename);
            sw.WriteLine("value");
            foreach (var item in list) sw.WriteLine(item);
        }

        static List<object> ImportList(string filename)
        {
            string content = File.ReadAllText(filename);
            if (filename.EndsWith(".json"))
                return JsonSerializer.Deserialize<List<object>>(content);
            else
            {
                var lines = content.Split('\n', StringSplitOptions.RemoveEmptyEntries);
                var result = new List<object>();
                for (int i = 1; i < lines.Length; i++)
                {
                    string line = lines[i].Trim();
                    if (int.TryParse(line, out int v)) result.Add(v);
                    else if (double.TryParse(line, out double d)) result.Add(d);
                    else result.Add(line);
                }
                return result;
            }
        }

        static List<object> GenerateRandomList(int size, int minVal, int maxVal)
        {
            var rnd = new Random();
            var list = new List<object>();
            for (int i = 0; i < size; i++)
                list.Add(rnd.Next(minVal, maxVal + 1));
            return list;
        }

        static async Task Interactive()
        {
            Console.WriteLine("🎲 ПЕРЕМЕШИВАТЕЛЬ СПИСКОВ");
            Console.WriteLine("Введите список (через запятую):");
            Console.Write("> ");
            string input = Console.ReadLine();
            List<object> data;
            if (string.IsNullOrWhiteSpace(input))
            {
                Console.Write("Размер случайного списка (по умолчанию 10): ");
                string sizeStr = Console.ReadLine();
                int size = string.IsNullOrEmpty(sizeStr) ? 10 : int.Parse(sizeStr);
                data = GenerateRandomList(size, 0, 100);
                Console.WriteLine($"Сгенерированный список: {FormatList(data)}");
            }
            else
            {
                data = ParseList(input);
                Console.WriteLine($"Исходный список: {FormatList(data)}");
            }

            Console.WriteLine("\nВыберите алгоритм:");
            foreach (var kv in algorithms)
                Console.WriteLine($"{kv.Key}. {kv.Value.name}");
            Console.Write("Ваш выбор (по умолчанию 1): ");
            string algoChoice = Console.ReadLine();
            if (string.IsNullOrEmpty(algoChoice) || !algorithms.ContainsKey(algoChoice)) algoChoice = "1";
            var algo = algorithms[algoChoice];

            Console.Write("Введите seed (или оставьте пустым для случайного): ");
            string seedStr = Console.ReadLine();
            Random rng;
            if (!string.IsNullOrEmpty(seedStr) && long.TryParse(seedStr, out long seed))
                rng = new Random((int)seed);
            else
                rng = new Random();

            var shuffled = algo.func(data, rng);
            Console.WriteLine($"\nПеремешанный список: {FormatList(shuffled)}");

            var stats = GetStats(shuffled);
            Console.WriteLine("\n📊 Статистика:");
            Console.WriteLine($"  Длина: {stats.length}");
            Console.WriteLine($"  Инверсий: {stats.inversions}");
            Console.WriteLine($"  Энтропия: {stats.entropy:F4}");

            Console.Write("\nЭкспортировать результат? (y/n): ");
            if (Console.ReadLine()?.ToLower() == "y")
            {
                Console.Write("Формат (json/csv): ");
                string fmt = Console.ReadLine();
                Console.Write("Имя файла: ");
                string filename = Console.ReadLine();
                if (string.IsNullOrEmpty(filename)) filename = $"shuffled.{fmt}";
                if (fmt == "json") ExportJSON(shuffled, filename);
                else ExportCSV(shuffled, filename);
                Console.WriteLine($"Экспортировано в {filename}");
            }
        }

        static void Main(string[] args)
        {
            if (args.Length > 0)
            {
                string list = null, file = null, algorithm = "1", export = null;
                int randomSize = 0;
                long seed = 0;
                bool statsFlag = false;
                for (int i = 0; i < args.Length; i++)
                {
                    if (args[i] == "--list") list = args[++i];
                    else if (args[i] == "--file") file = args[++i];
                    else if (args[i] == "--random") randomSize = int.Parse(args[++i]);
                    else if (args[i] == "--algorithm") algorithm = args[++i];
                    else if (args[i] == "--seed") seed = long.Parse(args[++i]);
                    else if (args[i] == "--export") export = args[++i];
                    else if (args[i] == "--stats") statsFlag = true;
                }
                List<object> data;
                if (list != null) data = ParseList(list);
                else if (file != null) data = ImportList(file);
                else if (randomSize > 0) data = GenerateRandomList(randomSize, 0, 100);
                else { Console.WriteLine("Укажите --list, --file или --random"); return; }
                if (data.Count == 0) { Console.WriteLine("Список пуст."); return; }

                Random rng = seed != 0 ? new Random((int)seed) : new Random();
                var algo = algorithms.ContainsKey(algorithm) ? algorithms[algorithm] : algorithms["1"];
                var shuffled = algo.func(data, rng);

                Console.WriteLine($"Исходный список: {FormatList(data)}");
                Console.WriteLine($"Перемешанный список: {FormatList(shuffled)}");
                if (statsFlag)
                {
                    var stats = GetStats(shuffled);
                    Console.WriteLine("\n📊 Статистика:");
                    Console.WriteLine($"  Длина: {stats.length}");
                    Console.WriteLine($"  Инверсий: {stats.inversions}");
                    Console.WriteLine($"  Энтропия: {stats.entropy:F4}");
                }
                if (export != null)
                {
                    if (export.EndsWith(".json")) ExportJSON(shuffled, export);
                    else ExportCSV(shuffled, export);
                    Console.WriteLine($"Экспортировано в {export}");
                }
            }
            else
            {
                Interactive().Wait();
            }
        }
    }
}
